From 60bd086d0da56bf26690e941b41b2a9fff19e1c8 Mon Sep 17 00:00:00 2001
From: Peter Harliman Liem <peter.harliman.liem@intel.com>
Date: Fri, 27 Mar 2020 15:13:46 +0800
Subject: [PATCH] PONRTSYS-6960 - Move spi nand device structure to use
 internal

Instead of directly defining flash properties with standard
nand_flash_dev struct, we switch it to use internal struct.
This is done to allow us extend with new properties when
needed, without us having to hack kernel header (nand_flash_dev).

This is a preparation work to add support for die-selection,
which will be done in upcoming commits.
---
 drivers/mtd/ltq-spinand/ltq_spinand.c | 80 +++++++++++++++++++++++------------
 1 file changed, 52 insertions(+), 28 deletions(-)

diff --git a/drivers/mtd/ltq-spinand/ltq_spinand.c b/drivers/mtd/ltq-spinand/ltq_spinand.c
index 34288d3f73c6..ab280ae50ed0 100644
--- a/drivers/mtd/ltq-spinand/ltq_spinand.c
+++ b/drivers/mtd/ltq-spinand/ltq_spinand.c
@@ -24,6 +24,12 @@
 #include "ltq_spinand.h"
 #include <linux/platform_device.h>
 
+struct spinand_flash_dev {
+	struct nand_flash_dev nand;
+
+	/* extension to standard nand flash structure */
+};
+
 /*============================================================================*/
 /* Macro and Switches */
 /*============================================================================*/
@@ -90,57 +96,76 @@ static inline struct spinand_info* spi_to_info(struct spi_device *spi)
  * If page size and eraseblock size are 0, the sizes are taken from the
  * extended chip ID.
  */
-struct nand_flash_dev spinand_flash_ids[] = {
-
-	{"SPI NAND Gigadevice 1Gbit 3.3v",
+static struct spinand_flash_dev spinand_flash_ids[] = {
+	{ {"SPI NAND Gigadevice 1Gbit 3.3v",
 		{ .id = {0xc8, 0xb1} }, 2048, 128, 0x20000,
 		 LP_OPTIONS | GIGADEVICE_DUMMY_TYPE, 2, 128},
-	{"SPI NAND Gigadevice 2Gbit 3.3v",
+		},
+	{ {"SPI NAND Gigadevice 2Gbit 3.3v",
 		{ .id = {0xc8, 0xb2} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | GIGADEVICE_DUMMY_TYPE, 2, 128},
-	{"SPI NAND Gigadevice 4Gbit 3.3v",
+		},
+	{ {"SPI NAND Gigadevice 4Gbit 3.3v",
 		{ .id = {0xc8, 0xb4} }, 4096, 512, 0x40000,
 		 LP_OPTIONS | GIGADEVICE_DUMMY_TYPE, 2, 256},
-	{"SPI NAND Gigadevice 4Gbit 1.8v",
+		},
+	{ {"SPI NAND Gigadevice 4Gbit 1.8v",
 		{ .id = {0xc8, 0xA4} }, 4096, 512, 0x40000,
 		 LP_OPTIONS | GIGADEVICE_DUMMY_TYPE, 2, 256},
-	{"SPI NAND Micron 128MiB 3,3V 8-bit",
+		},
+	{ {"SPI NAND Micron 128MiB 3,3V 8-bit",
 		{ .id = {0x2c, 0x11} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND Micron 128MiB 3,3V 8-bit",
+		},
+	{ {"SPI NAND Micron 128MiB 3,3V 8-bit",
 		{ .id = {0x2c, 0x12} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND MT29F2G01AAAED 256MiB",
+		},
+	{ {"SPI NAND MT29F2G01AAAED 256MiB",
 		{ .id = {0x2c, 0x22} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 64},
-	{"SPI NAND MT29F1G01ABAFD 128MiB",
+		},
+	{ {"SPI NAND MT29F1G01ABAFD 128MiB",
 		{ .id = {0x2c, 0x14} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 128},
-	{"SPI NAND MT29F2G01ABAGD 256MiB",
+		},
+	{ {"SPI NAND MT29F2G01ABAGD 256MiB",
 		{ .id = {0x2c, 0x24} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 128},
-	{"SPI NAND MT29F4G01AAADD 512MiB",
+		},
+	{ {"SPI NAND MT29F4G01AAADD 512MiB",
 		{ .id = {0x2c, 0x32} }, 2048, 512, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 64},
-	{"SPI NAND MT29F1G01ABB 1Gbit",
+		},
+	{ {"SPI NAND MT29F1G01ABB 1Gbit",
 		{ .id = {0x2c, 0x15} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 128},
-	{"SPI NAND MT29F2G01ABBG 2Gbit 1,8v",
+		},
+	{ {"SPI NAND MT29F2G01ABBG 2Gbit 1,8v",
 		{ .id = {0x2c, 0x25} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 64},
-	{"SPI NAND TC58CVG0S3 1Gbit",
+		},
+	{ {"SPI NAND TC58CVG0S3 1Gbit",
 		{ .id = {0x98, 0xc2} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND TC58CVG1S3 2Gbit",
+		},
+	{ {"SPI NAND TC58CVG1S3 2Gbit",
 		{ .id = {0x98, 0xcb} }, 2048, 256, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND TC58CVG2S0 4Gbit",
+		},
+	{ {"SPI NAND TC58CVG2S0 4Gbit",
 		{ .id = {0x98, 0xcd} }, 4096, 512, 0x40000, LP_OPTIONS, 2, 128},
-	{"SPI NAND MX35LFE4AB 1Gbit",
+		},
+	{ {"SPI NAND MX35LFE4AB 1Gbit",
 		{ .id = {0xc2, 0x12} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND MX35LFE4AB 2Gbit",
+		},
+	{ {"SPI NAND MX35LFE4AB 2Gbit",
 		{ .id = {0xc2, 0x22} }, 2048, 256, 0x20000,
 		 LP_OPTIONS | SPINAND_NEED_PLANE_SELECT, 2, 64},
-	{"SPI NAND 1Gbit 3,3v WN25N01GV",
+		},
+	{ {"SPI NAND 1Gbit 3,3v WN25N01GV",
 		{ .id = {0xef, 0xaa} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND 1Gbit 1,8v WN25N01GW",
+		},
+	{ {"SPI NAND 1Gbit 1,8v WN25N01GW",
 		{ .id = {0xef, 0xba} }, 2048, 128, 0x20000, LP_OPTIONS, 2, 64},
-	{"SPI NAND 2Gbit 3,3v WN25M02GV",
+		},
+	{ {"SPI NAND 2Gbit 3,3v WN25M02GV",
 		{ .id = {0xef, 0xab} }, 2048, 256, 0x20000, LP_OPTIONS, 2, 64},
+		},
 };
 #if 0
 /**
@@ -328,20 +353,19 @@ struct nand_flash_dev * spinand_flash_detect(struct mtd_info *mtd, struct nand_c
 {
 	u8 id_data[8];
 	int i;
-	struct nand_flash_dev *type = NULL;
+	struct nand_flash_dev *type;
 
 	chip->cmdfunc(mtd, NAND_CMD_READID, 0x00, -1);
 	for (i = 0; i < 3; i++)
 		id_data[i] = chip->read_byte(mtd);
 
-	if (!type)
-		type = spinand_flash_ids;
-
-	for (; type->name != NULL; type++)
+	for (i = 0; i < ARRAY_SIZE(spinand_flash_ids); i++) {
+		type = &spinand_flash_ids[i].nand;
 		if ((id_data[0] == type->mfr_id) && (id_data[1] == type->dev_id))
 			break;
+	}
 
-	if (!type->name) {
+	if (i == ARRAY_SIZE(spinand_flash_ids)) {
 		pr_err("Unable to find flash id 0x%x 0x%x\n", id_data[0],
 		       id_data[1]);
 		goto probe_done;
